examples/jsm/loaders/MaterialXLoader.js

```diff
@@ -0,0 +1,728 @@
+
+				}
+
+			}, onProgress, onError );
+
+		return this;
+
+	}
+
+	parse( text ) {
+
+		return new MaterialX( this.manager, this.path ).parse( text );
+
+	}
+
+}
+
+class MaterialXNode {
+
+	constructor( materialX, nodeXML, nodePath = '' ) {
+
+		this.materialX = materialX;
+		this.nodeXML = nodeXML;
+		this.nodePath = nodePath ? nodePath + '/' + this.name : this.name;
+
+		this.parent = null;
+
+		this.node = null;
+
+		this.children = [];
+
+	}
+
+	get element() {
+
+		return this.nodeXML.nodeName;
+
+	}
+
+	get nodeGraph() {
+
+		return this.getAttribute( 'nodegraph' );
+
+	}
+
+	get nodeName() {
+
+		return this.getAttribute( 'nodename' );
+
+	}
+
+	get interfaceName() {
+
+		return this.getAttribute( 'interfacename' );
+
+	}
+
+	get output() {
+
+		return this.getAttribute( 'output' );
+
+	}
+
+	get name() {
+
+		return this.getAttribute( 'name' );
+
+	}
+
+	get type() {
+
+		return this.getAttribute( 'type' );
+
+	}
+
+	get value() {
+
+		return this.getAttribute( 'value' );
+
+	}
+
+	getNodeGraph() {
+
+		let nodeX = this;
+
+		while ( nodeX !== null ) {
+
+			if ( nodeX.element === 'nodegraph' ) {
+
+				break;
+
+			}
+
+			nodeX = nodeX.parent;
+
+		}
+
+		return nodeX;
+
+	}
+
+	getRoot() {
+
+		let nodeX = this;
+
+		while ( nodeX.parent !== null ) {
+
+			nodeX = nodeX.parent;
+
+		}
+
+		return nodeX;
+
+	}
+
+	get referencePath() {
+
+		let referencePath = null;
+
+		if ( this.nodeGraph !== null && this.output !== null ) {
+
+			referencePath = this.nodeGraph + '/' + this.output;
+
+		} else if ( this.nodeName !== null || this.interfaceName !== null ) {
+
+			referencePath = this.getNodeGraph().nodePath + '/' + ( this.nodeName || this.interfaceName );
+
+		}
+
+		return referencePath;
+
+	}
+
+	get hasReference() {
+
+		return this.referencePath !== null;
+
+	}
+
+	get isConst() {
+
+		return this.element === 'input' && this.value !== null && this.type !== 'filename';
+
+	}
+
+	getColorSpaceNode() {
+
+		const csSource = this.getAttribute( 'colorspace' );
+		const csTarget = this.getRoot().getAttribute( 'colorspace' );
+
+		const nodeName = `mx_${ csSource }_to_${ csTarget }`;
+
+		return colorSpaceLib[ nodeName ];
+
+	}
+
+	getTexture() {
+
+		const filePrefix = this.getRecursiveAttribute( 'fileprefix' ) || '';
+
+		const texture = this.materialX.textureLoader.load( filePrefix + this.value );
+		texture.wrapS = texture.wrapT = RepeatWrapping;
+		texture.flipY = false;
+
+		return texture;
+
+	}
+
+	getClassFromType( type ) {
+
+		let nodeClass = null;
+
+		if ( type === 'integer' ) nodeClass = int;
+		else if ( type === 'float' ) nodeClass = float;
+		else if ( type === 'vector2' ) nodeClass = vec2;
+		else if ( type === 'vector3' ) nodeClass = vec3;
+		else if ( type === 'vector4' || type === 'color4' ) nodeClass = vec4;
+		else if ( type === 'color3' ) nodeClass = color;
+		else if ( type === 'boolean' ) nodeClass = bool;
+
+		return nodeClass;
+
+	}
+
+	getNode() {
+
+		let node = this.node;
+
+		if ( node === null ) {
```

Comment:
Maybe `if ( node !== null ) { return node; }`? This will decrease complexity of the remaining code by one level and will make it easier to read.